agent/src/main/java/org/openremote/agent/protocol/modbus/ModbusTcpProtocol.java

```diff
@@ -19,36 +19,123 @@
-    }
+public class ModbusTcpProtocol extends AbstractModbusProtocol<ModbusTcpProtocol, ModbusTcpAgent> {
+
+    public static final Logger LOG = SyslogCategory.getLogger(SyslogCategory.PROTOCOL, ModbusTcpProtocol.class);
+    private ModbusTcpIOClient client = null;
+    private final Map<Integer, CompletableFuture<ModbusTcpFrame>> pendingRequests = new ConcurrentHashMap<>();
+    private final Map<Integer, Long> timedOutRequests = new ConcurrentHashMap<>(); // Track timed-out TxIDs with timestamp
 
-    private String connectionString;
+    public ModbusTcpProtocol(ModbusTcpAgent agent) {super(agent);}
 
     @Override
-    protected PlcConnection createIoClient(ModbusTcpAgent agent) throws RuntimeException {
-        PlcConnection plcConnection;
-        connectionString = "modbus-tcp://" + agent.getHost().orElseThrow() + ":" + agent.getPort().orElseThrow()+"?unit-identifier=" + agent.getUnitId();
+    protected void doStartProtocol(Container container) throws Exception {
+        String host = agent.getHost().orElseThrow(() -> new IllegalStateException("Host not configured"));
+        int port = agent.getPort().orElseThrow(() -> new IllegalStateException("Port not configured"));
+        connectionString = "modbus-tcp://" + host + ":" + port;
+
         try {
-            plcConnection = PlcDriverManager.getDefault().getConnectionManager().getConnection(connectionString);
-        } catch (PlcConnectionException e) {
-            throw new RuntimeException(e);
+            client = new ModbusTcpIOClient(host, port);
+            client.addMessageConsumer(this::handleIncomingFrame);
+            client.addConnectionStatusConsumer(this::setConnectionStatus);
+            client.connect();
+            LOG.info("Modbus TCP connection initiated for " + connectionString);
+        } catch (Exception e) {
+            LOG.log(Level.SEVERE, "Failed to create Modbus TCP client: " + agent, e);
+            setConnectionStatus(ConnectionStatus.ERROR);
+            throw e;
         }
-        return plcConnection;
     }
 
     @Override
-    public String getProtocolName() {
-        return "Modbus TCP Protocol";
+    protected void doStopProtocol(Container container) throws Exception {
+        if (client != null) {
+            client.disconnect();
+            client = null;
+        }
+        pendingRequests.clear();
+    }
+
+    @Override
+    protected Optional<ModbusAgent.DeviceConfigMap> getDeviceConfig() {
+        return agent.getDeviceConfig();
+    }
+
+    @Override
+    protected ModbusResponse sendModbusRequest(int unitId, byte[] pdu, long timeoutMs) throws Exception {
+        return sendRequestAndWaitForResponse(unitId, pdu, timeoutMs);
     }
 
     @Override
-    public String getProtocolInstanceUri() {
-        return connectionString;
+    public String getProtocolName() {
+        return "TCP";
+    }
+
+    private ModbusTcpFrame sendRequestAndWaitForResponse(int unitId, byte[] pdu, long timeoutMs) throws Exception {
```

Comment:
Consider using protected accessors to allow easier subclassing of this class in future